spec/support/shared_examples_api_request.rb

```diff
@@ -0,0 +1,343 @@
+#
+#  These are the parameters passed to this example:
+#    item_class [Class] - the class for the item; is used to create a new item
+#      with the item_params, and to get the slug for the created item.
+#    item_name [String] - the name of the item (singular).
+#      It is used as a key in the parameters sent to the server
+#      and as part of the schema name in the schema definitions.
+#      It is pluralized and used for the path and the tags.
+#
+RSpec.shared_examples 'it creates an item for a user' do |item_class, item_name|
+  item_plural = item_name.pluralize
+  item_name_symbol = item_name.to_sym
+
+  path "#{VALID_API_PATH}/#{item_plural}" do
+    post "creates a #{item_name}" do
+      tags item_plural
+      consumes 'application/json', 'application/xml'
+      parameter name: 'Accept', in: :header, type: :string, default: 'application/vnd.pieforproviders.v1+json'
+      parameter name: item_name_symbol, in: :body, schema: {
+        '$ref' => "#/definitions/create#{item_class}"
+      }
+
+      context 'on the right api version' do
+        include_context 'correct api version header'
+        context 'when authenticated' do
+          include_context 'authenticated user'
+
+          response '201', "#{item_name} created" do
+            let(item_name_symbol) { { item_name => item_params } }
+            run_test! do
+              expect(response).to match_response_schema(item_name)
+            end
+          end
+
+          response '422', 'invalid request' do
+            let(item_name_symbol) { { item_name => { 'blorf': 'whatever' } } }
+            run_test!
+          end
+        end
+
+        it_behaves_like '401 error if not authenticated with parameters', item_name
+      end
+
+      it_behaves_like 'server error responses for wrong api version with parameters', item_name
+    end
+  end
+end
+
+# This example expects the following to be defined with a let(:) block:
+#  item_params - parameters to be passed to the server
+#
+#  These are the parameters passed to this example:
+#    item_class [Class] - the class for the item; is used to create a new item
+#      with the item_params, and to get the slug for the created item.
+#    item_name [String] - the name of the item (singular).
+#      It is used as a key in the parameters sent to the server
+#      and as part of the schema name in the schema definitions.
+#      It is pluralized and used for the path and the tags.
+#    update_attribute [String] - attribute name to be updated
+#    update_valid_value [String | Number | nil] - valid value for the updated value for the attribute
+#    update_invalid_value  [String | Number | nil] - invalid value for the attribute so that the server returns a 422 (cannot be updated) error
+#
+RSpec.shared_examples 'it updates an item with a slug, for a user' do |item_class, item_name, update_attribute, update_valid_value, update_invalid_value|
+  item_plural = item_name.pluralize
```

Comment:
```suggestion
  item_name = item_class.class.name.downcase
  item_plural = item_name.pluralize
```

Reply:
I'd love to see this instead of the item name passed to every one of these shared examples.  I didn't mark them all up, just the first few examples.

Reply:
I definitely thought about this too.  Didn't know how much I could really assume about the item name: would it always be based on the class name?  I couldn't think of a situation where it would _not_ be, but didn't really know.
Since you've confirmed that we can assume that -- I'll definitely make the change.  Simplifies things. 👍